### PR TITLE
Adding typescript definition for 'this' on Middleware and Afterware apply functions.

### PR DESCRIPTION
**Pull request by vangorra**
- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [x] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ Expect active development and potentially significant breaking changes in the `0
 - Fix: make sure maybeDeepFreeze is called on results returned from setVariables and refetch [PR #1362](https://github.com/apollographql/apollo-client/pull/1362)
 - Fix: use setTimeout to throw uncaught errors in observer.next and observer.error[PR #1367](https://github.com/apollographql/apollo-client/pull/1367)
 - Remove returnPartialData option [PR #1370](https://github.com/apollographql/apollo-client/pull/1370)
-
+- Update TypeScript Middleware and Afterware interfaces to include a datatype for 'this' in apply function.
```
Comment (helfer) [inline CHANGELOG.md]: Make sure to add the PR number here!

File: src/transport/afterware.ts [hunk 2]
```diff
@@ -13,5 +15,5 @@ export interface BatchAfterwareResponse {
 }
 
 export interface BatchAfterwareInterface {
-  applyBatchAfterware(response: BatchAfterwareResponse, next: Function): any;
+  applyBatchAfterware(this: NetworkInterface, response: BatchAfterwareResponse, next: Function): any;
```
Comment (helfer) [inline src/transport/afterware.ts]: Is this the correct type or could it be more specific (BatchNetworkInterface)?

File: src/transport/middleware.ts [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
+import { Request, NetworkInterface } from './networkInterface';
 
 export interface MiddlewareRequest {
   request: Request;
   options: RequestInit;
 }
 
 export interface MiddlewareInterface {
-  applyMiddleware(request: MiddlewareRequest, next: Function): void;
+  applyMiddleware(this: NetworkInterface, request: MiddlewareRequest, next: Function): void;
```
Comment (helfer) [inline src/transport/middleware.ts]: Also this type. Could it be more specific?

**Comment by stubailo** [comment]
Looks great, thanks!

**Pull request merged**